src/hotspot/share/cds/archiveBuilder.cpp

```diff
@@ -1037,44 +1039,162 @@ class ArchiveBuilder::CDSMapLogger : AllStatic {
+        // This is a regular oop that got archived.
+        print_oop_with_requested_addr_cr(&st, source_oop, false);
+        byte_size = source_oop->size() * BytesPerWord;
       } else if (start == ArchiveHeapWriter::buffered_heap_roots_addr()) {
-        // HeapShared::roots() is copied specially so it doesn't exist in
-        // HeapShared::OriginalObjectTable. See HeapShared::copy_roots().
-        log_info(cds, map)(PTR_FORMAT ": @@ Object HeapShared::roots (ObjArray)",
-                           p2i(to_requested(start)));
+        // HeapShared::roots() is copied specially, so it doesn't exist in
+        // ArchiveHeapWriter::BufferOffsetToSourceObjectTable.
+        // See ArchiveHeapWriter::copy_roots_to_buffer().
+        st.print_cr("HeapShared::roots[%d]", HeapShared::pending_roots()->length());
         byte_size = ArchiveHeapWriter::heap_roots_word_size() * BytesPerWord;
+      } else if ((byte_size = ArchiveHeapWriter::get_filler_size_at(start)) > 0) {
+        // We have a filler oop, which also does not exist in BufferOffsetToSourceObjectTable.
+        st.print_cr("filler " SIZE_FORMAT " bytes", byte_size);
       } else {
-        // We have reached the end of the region, but have some unused space
-        // at the end.
-        log_info(cds, map)(PTR_FORMAT ": @@ Unused heap space " SIZE_FORMAT " bytes",
-                           p2i(to_requested(start)), size_t(end - start));
-        log_data(start, end, to_requested(start), /*is_heap=*/true);
-        break;
+        ShouldNotReachHere();
       }
+
       address oop_end = start + byte_size;
-      log_data(start, oop_end, to_requested(start), /*is_heap=*/true);
+      log_as_hex(start, oop_end, requested_start, /*is_heap=*/true);
+
+      if (source_oop != nullptr) {
+        log_oop_details(heap_info, source_oop);
+      } else if (start == ArchiveHeapWriter::buffered_heap_roots_addr()) {
+        log_heap_roots();
+      }
       start = oop_end;
     }
   }
 
-  static address to_requested(address p) {
-    return ArchiveHeapWriter::buffered_addr_to_requested_addr(p);
+  // ArchivedFieldPrinter is used to print the fields of archived objects. We can't
+  // use _source_obj->print_on(), because we want to print the oop fields
+  // in _source_obj with their requested addresses using print_oop_with_requested_addr_cr().
+  class ArchivedFieldPrinter : public FieldClosure {
+    ArchiveHeapInfo* _heap_info;
+    outputStream* _st;
+    oop _source_obj;
+  public:
+    ArchivedFieldPrinter(ArchiveHeapInfo* heap_info, outputStream* st, oop src_obj) :
+      _heap_info(heap_info), _st(st), _source_obj(src_obj) {}
+    void do_field(fieldDescriptor* fd) {
+      _st->print(" - ");
+      BasicType ft = fd->field_type();
+      switch (ft) {
+      case T_ARRAY:
+      case T_OBJECT:
+        fd->print_on(_st); // print just the name and offset
+        print_oop_with_requested_addr_cr(_st, _source_obj->obj_field(fd->offset()));
+        break;
+      default:
```

Comment:
The if block inside the default case leads to a lot of indentation. Maybe this could be it's own method?